### PR TITLE
Fix: Docker Daemon didn't send actual actions client ask for to Token Server

### PR DESCRIPTION
**Pull request by xiekeyang**
daemon commit pull and push action values for authorization on pull session,
that seems incorrect.

So I fix it, to use pull action on pull session, and use pull and push
actions on push session.

Such bug on api-trust.

Signed-off-by: xiekeyang xiekeyang@huawei.com

### CODE REVIEW

File: api/client/trust.go [hunk 1]
```diff
@@ -156,7 +156,7 @@ func (cli *DockerCli) getNotaryRepository(repoInfo *registry.RepositoryInfo, aut
 	}
 
 	creds := simpleCredentialStore{auth: authConfig}
-	tokenHandler := auth.NewTokenHandler(authTransport, creds, repoInfo.CanonicalName, "push", "pull")
+	tokenHandler := auth.NewTokenHandler(authTransport, creds, repoInfo.CanonicalName, []string{"push", "pull"})
```
Comment (endophage) [inline api/client/trust.go]: No need to wrap in a `[]string{}`, the last argument to `auth.NewTokenHandler` is `actions ...string`

https://github.com/docker/distribution/blob/master/registry/client/auth/session.go#L113

Comment (endophage) [inline api/client/trust.go]: p.s. Don't make `getNotaryRepository` take a `...string`, it's still early days for trust and that would potentially make changes harder.

File: vendor/src/github.com/docker/distribution/registry/client/auth/session.go [hunk 2]
```diff
@@ -110,7 +110,7 @@ func (ts tokenScope) String() string {
 
 // NewTokenHandler creates a new AuthenicationHandler which supports
 // fetching tokens from a remote token server.
-func NewTokenHandler(transport http.RoundTripper, creds CredentialStore, scope string, actions ...string) AuthenticationHandler {
+func NewTokenHandler(transport http.RoundTripper, creds CredentialStore, scope string, actions []string) AuthenticationHandler {
```
Comment (endophage) [inline vendor/src/github.com/docker/distribution/registry/client/auth/session.go]: Unnecessary change that may have a wider impact. Would prefer this is taken out.

(Also, this is not the appropriate place to make this change, anything in /vendor needs to be changed at the source).

**Comment by endophage** [comment]
This is sensible cleanup. Priority wise, not sure I'd classify it as a bug. 

The token server will only issue tokens for the permissions you actually have, not all the permissions that you ask for. This may mean you have more permissions than necessary for a pull, but that wont break anything.

File: vendor/src/github.com/docker/distribution/registry/client/auth/session.go (hunk 2 above)
Comment (stevvooe) [inline vendor/src/github.com/docker/distribution/registry/client/auth/session.go]: I see no advantage to making this change.

File: graph/registry.go [hunk 3]
```diff
@@ -91,7 +91,7 @@ func NewV2Repository(repoInfo *registry.RepositoryInfo, endpoint registry.APIEnd
 	}
 
 	creds := dumbCredentialStore{auth: authConfig}
-	tokenHandler := auth.NewTokenHandler(authTransport, creds, repoName, "push", "pull")
+	tokenHandler := auth.NewTokenHandler(authTransport, creds, repoName, actions)
```
Comment (stevvooe) [inline graph/registry.go]: This can be `tokenHandler := auth.NewTokenHandler(authTransport, creds, repoName, actions...)`. No need to change function signature.

**Comment by xiekeyang** [comment]
@endophage  and @stevvooe ,
I change my PR, to fix the actual actions only in graph/registry, not including trust mod. and use the parameters type of ...string . Please take a look. Thanks a lot!

**Comment by stevvooe** [comment]
@dmcgowan Shouldn't `NewTokenHandler` resolve and request the appropriate scope?

**Comment by dmcgowan** [comment]
I really don't have an opinion on whether it should or shouldn't. David is right that "This may mean you have more permissions than necessary for a pull, but that wont break anything". While I think keeping scope small is ideal, if we were to switch to a longer term token model I think it will be better to have the maximum scope. However the token server "should" support the additional scopes, so it is not a bug and not "necessary". My opinion is we should avoid this sort of unnecessary change.

**Comment by xiekeyang** [comment]
@dmcgowan The scope request is sent to authenticate server which `issuer` defined. auth server should know the appropriate scope, and then decide if the `account` has the permission to `scope actions`. if daemon  don't send the appropriate `scope actions`, for example:
* `account` want to pull a public image which can only be pulled by users, but daemon send `pull` and `push` actions to auth server.
* server will regard the actions as `push`, and denied the request.
without `scope`, server is hard to authenticate user's behavior.
I learn some auth server design, that use scope action as a necessary condition. I'm not sure if it is a bug. But it really impact on `pull` and `push` session authorization.
So could you please take my comment into consideration? Thanks a lot!

**Comment by calavera** [comment]
I agree with @xiekeyang that we should request only auth scopes for the action we want to perform.

LGTM.

**Comment by stevvooe** [comment]
@calavera Part of the issue here is that the token contract is between the issuer and the consumer (in this case, the registry). The client asks the issuer for the permissions and then passes the opaque token to the registry, at which time it may find that the required permissions are not available. The client has no ability, nor should it have said ability, to validate whether the permissions are sufficient until they are tested against an authority.

Either way, this model of passing in the required tokens at transport creation is slightly broken. The permissions should be detected as requests come through the transport and they should be resolved at that time, eliding this extra complexity. This also lets the client decide how it wants to manage permissions. For example, a careful client can build up the token over several requests, whereas other extremes may request a token for each request or request all permissions at once.

The actual bug here is in non-compliant token endpoints that incorrectly return an error when a certain level of permissions is not available. A token with empty scopes is just as valid as a token with all scopes but it is up to the registry to make the decision about validity. I'm not sure if @jlhawn clarified this in the specification. Effectively, the decision as to whether a token provides the client the requested operation is completely the responsibility of the registry's authority.

**Comment by endophage** [comment]
@calavera I agree with you. There a question that needs to be answered around the token server however: what should the behaviour be when a user requests permissions they don't have? Our current implementation will issue a token with only the permissions the user actually has, so while it's cleaner to request only the permissions needed, no error is raised by over-requesting.

I believe others (@xiekeyang?) that have written their own token servers, have designed them such that the token server will refuse to issue a token if it can't satisfy all the permissions requested.

This discussion should be taken elsewhere (@stevvooe: do we have an appropriate place to discuss and/or document our expectations for a token server?). 

As I said above, this is good, sensible cleanup so I'm happy to give my LGTM now that the issues identified earlier have been addressed.

**Comment by stevvooe** [comment]
@endophage Do I agree with @calavera in the principle that "we should request only auth scopes for the action we want to perform"? Yes. However, we must temper this with other considerations.

This is less of a discussion than a description of the contract for obtaining tokens and the responsibility of validating them. With the described model above, well-implemented clients will only have to comply with the token API. Everything else, including the permissions and policy error response codes, can remain opaque. If we adopt a strict model, where the authentication server is expected to meet the requested scopes and validate, it is effectively an API change, which is demonstrated by this PR. Not only does it change the expectation of the responses from the token server, but it also requires us to define a policy error reporting mechanism in the token API, in addition to that already provided by the registry API.

The current model allows us to opaquely swap out the token scheme without changing the clients implementation, as long as the client keeps track of which tokens provided which permissions.

Conversely, an implementation may preemptively return all of the token permissions in a single token for the user, if it deems it necessary. It may not yet know it has all those permissions until it requests them from the token server, but this should be okay. Let's take the situation where a fat token is issued, then unused permissions are taken away. Would it be an error to take away permissions from a token that already had a set of permissions that were never originally requested by the client? In the current model, this is trivially implemented in a backwards compatible manner. With a strict model, the correct behavior is less apparent.

The correct answer to this PR would be for the implementor of the authentication server to issue the tokens without the requested scope, rather than an error. This condition is not described in the specification as an error and implementations that return an error under these conditions are simply not in compliance. It is rather unfortunate that the original implementation was more careful about the requested permissions, making this lack of clarity in the specification unapparent until now. We can be flexible here, but considering this is a new protocol, there is no reason these new implementations shouldn't comply with the original interpretation.

I know understanding why this is all related is subtle, but if we look at the token acquisition process, it becomes somewhat apparent. Specifically, during the course of a registry bound request, we may want to request a new or updated token. If the auth endpoint may return an error that the client doesn't have necessary scopes, we must return an http error for a request that never egressed to the target endpoint. While this is a problem in Go's `net/http` package, it is also a question of policy error handling. __The auth server doesn't have an API to describe policy errors and allow a client to resolve them. This is part of the registry API or whatever service has the responsibility to validate the tokens.__

__TL; DR__ LGTM with the acknowledgement that @xiekeyang's auth token server implementation is not compliant.

**Comment by xiekeyang** [comment]
@stevvooe 
Thanks a lot for your explain. I agree with you that auth server should provide exact permission. So, I would cancel this PR.
Then, the actions query in `Requesting a Token` API becomes invalid.
```bash
GET /v2/token/?service=registry.docker.com&scope=repository:samalba/my-app:push&account=jlhawn
```
Do you think we should remove it from API, and update related spec?
Thanks a lot!

**Comment by stevvooe** [comment]
@xiekeyang We are updating the token spec in https://github.com/docker/distribution/pull/973 to clarify this.

I don't see any problems with merging this PR.

LGTM

**Comment by thaJeztah** [comment]
@calavera @stevvooe I think we have two LGTM's here; is this ready to merge?

**Comment by calavera** [comment]
Yup, thank you all for the reviews and changes. :tada:
